zapisy/apps/schedule/forms.py

```diff
@@ -99,14 +99,58 @@ class CustomVisibleCheckbox(Field):
     template = 'forms/custom_visible_checkbox.html'
 
 
+# PretendRequired classes are designed to bypass the execution of the Field.validate method which
+# raises an exception if a required field is empty. This ensures that the field appears as
+# required to the user, but the default validation checks are omitted. We perform our own
+# validation inside the Event.clean method, which requires fields not to be empty only if certain
+# conditions are met.
+# For details on the CharField.validate and ModelChoiceField.validate methods, refer to:
+# https://github.com/django/django/blob/6726d750979a7c29e0dd866b4ea367eef7c8a420/django/forms/fields.py#L134
+# https://github.com/django/django/blob/6726d750979a7c29e0dd866b4ea367eef7c8a420/django/forms/models.py#L977
+class PretendRequiredCharField(forms.CharField):
+    def __init__(self, *args, **kwargs):
+        if not kwargs.get('required', True):
+            raise ValueError("The 'required' argument cannot be set to False")
+        super().__init__(*args, **kwargs)
+
+    def validate(self, value):
+        pass
+
+
+class PretendRequiredModelChoiceField(forms.ModelChoiceField):
+    def __init__(self, *args, **kwargs):
+        if not kwargs.get('required', True):
+            raise ValueError("The 'required' argument cannot be set to False")
+        super().__init__(*args, **kwargs)
+
+    def validate(self, value):
+        pass
+
+
 class EventForm(forms.ModelForm):
     class Meta:
         model = Event
         exclude = ('status', 'author', 'created', 'edited', 'group', 'interested')
 
-    title = forms.CharField(label="Nazwa", required=False)
+    def clean(self):
+        cleaned_data = super().clean()
+
+        if cleaned_data['type'] not in (Event.TYPE_EXAM, Event.TYPE_TEST):
+            cleaned_data['course'] = None
+        if cleaned_data['type'] != Event.TYPE_GENERIC:
+            cleaned_data['title'] = None
```

Comment:
To nie jest pytanie retoryczne: czy tak byłoby przejrzyściej, czy wręcz przeciwnie? Pewne testy przestają się powtarzać, ale to wcale nie musi być krok w dobrą stronę...
```suggestion
        if cleaned_data['type'] not in (Event.TYPE_EXAM, Event.TYPE_TEST):
            cleaned_data['course'] = None
        elif cleaned_data['course'] == None:
            raise # wyjątek
        
        if cleaned_data['type'] != Event.TYPE_GENERIC:
            cleaned_data['title'] = None
        elif cleaned_data['title'] in ('', None):
            raise # wyjątek
```

Reply:
Początkowo wydawało mi się, że wersja pierwotna jest trochę bardziej przejrzysta, bo bloki kodu (te oddzielone pustymi liniami) dotyczą odrębnych kwestii łatwiejszych do zauważenia. W pierwszym czyścimy niepotrzebne pola, a w drugim sprawdzamy, czy pola obowiązkowe są wypełnione. W zasugerowanej zmianie te bloki są podzielone względem tego, którym z pól `cleaned_data` się zajmujemy, co moim zdaniem jest trochę mniej czytelne. Przy dodawaniu kolejnych pól (np. `thesis`) będziemy dodawać kolejne bloki zamiast po jednym ifie do każdego z dwóch bloków. Może tak będzie lepiej a może nie, ale warto wziąć to pod uwagę. Na początku ten `elif ...` wydał mi się mało czytelny, w pierwotnej wersji można usunąć `else` a `elif` zamienić na `if` (to nie ja pisałem tamten fragment) i przejrzystość moim zdaniem jeszcze bardziej się poprawi. Ogólnie to im dłużej o tym myślę, tym bardziej odnoszę wrażenia, że czytelność obu wersji jest taka sama.